### ISSUE TITLE
MetricRegistry::counter inconsistent result with multiple tags

### ISSUE THREAD

**Issue by vgovtva** [helidon-io/helidon#5541]
## Environment Details
* Helidon Version: 3.0.*
* Helidon SE or Helidon MP: MP
* JDK version: openjdk 17.0.4 2022-07-19
* OS: Debian 11

----------

## Problem Description
`MetricRegistry::counter` does not return consistent result when you have more than 1 tag specified.
It seems that the order in which you provide these tags affects the behavior. If your tags are alphabetically sorted, then you get the same counter every time you call the method. But if you have tags in non-alphabetical order, then seemingly you get a new instance of a counter every time you call the method.
Expected Behavior: when calling `MetricRegsitry::counter` with some string and a set of tags, regardless of the order of tags, you should get the same counter object back.
```
Counter counter1 = metricRegistry.counter("myCounter" , new Tag("foo", "1"),  new Tag("bar", "1"));
Counter counter2 = metricRegistry.counter("myCounter" , new Tag("foo", "1"),  new Tag("bar", "1"));
Counter counter3 = metricRegistry.counter("myCounter" , new Tag("bar", "1"),  new Tag("foo", "1"));

// These will fail
assertEquals(counter1, counter2);
assertEquals(counter1, counter3);

// In practice
Counter counter4 = metricRegistry.counter("myCounter" , new Tag("bar", "1"),  new Tag("foo", "1"));
assertEquals(counter3, counter4);  // This will be true
```

## Steps to reproduce
Adding tests below makes build fail on my local machine. (These are not exactly `MetricRegsitry` Tests, but I believe it's part of the same bug?)
```
$ git --no-pager diff HEAD
diff --git a/metrics/api/src/test/java/io/helidon/metrics/api/MetricStoreTests.java b/metrics/api/src/test/java/io/helidon/metrics/api/MetricStoreTests.java
index d6930dc26..cbce25c63 100644
--- a/metrics/api/src/test/java/io/helidon/metrics/api/MetricStoreTests.java
+++ b/metrics/api/src/test/java/io/helidon/metrics/api/MetricStoreTests.java
@@ -22,6 +22,7 @@ import org.eclipse.microprofile.metrics.MetricType;
 import org.eclipse.microprofile.metrics.SimpleTimer;
 import org.eclipse.microprofile.metrics.Tag;
 import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;

 class MetricStoreTests {
@@ -56,4 +57,92 @@ class MetricStoreTests {
         assertThrows(IllegalArgumentException.class, () ->
                 store.getOrRegisterMetric(meta2, Counter.class, NO_TAGS));
     }
+
+    @Test
+    void testAvoidDuplicates() {
+        Metadata meta1 = Metadata.builder()
+                .withName("a")
+                .withType(MetricType.COUNTER)
+                .build();
+
+        NoOpMetricRegistry registry = NoOpMetricRegistry.create(MetricRegistry.Type.APPLICATION);
+
+        MetricStore<NoOpMetric> store = MetricStore.create(REGISTRY_SETTINGS,
+                                                           NoOpMetricRegistry.NO_OP_METRIC_FACTORIES,
+                                                           null,
+                                                           null,
+                                                           MetricRegistry.Type.APPLICATION,
+                                                           NoOpMetric.class,
+                                                           registry::toImpl);
+
+        Counter counter1 = store.getOrRegisterMetric(meta1, Counter.class, NO_TAGS);
+        Counter counter2 = store.getOrRegisterMetric(meta1, Counter.class, NO_TAGS);
+        assertEquals(counter1, counter2);
+    }
+    @Test
+    void testAvoidDuplicates1() {
+        Tag[] TAGS = {new Tag("foo", "1")};
+        Metadata meta1 = Metadata.builder()
+                .withName("a")
+                .withType(MetricType.COUNTER)
+                .build();
+
+        NoOpMetricRegistry registry = NoOpMetricRegistry.create(MetricRegistry.Type.APPLICATION);
+
+        MetricStore<NoOpMetric> store = MetricStore.create(REGISTRY_SETTINGS,
+                                                           NoOpMetricRegistry.NO_OP_METRIC_FACTORIES,
+                                                           null,
+                                                           null,
+                                                           MetricRegistry.Type.APPLICATION,
+                                                           NoOpMetric.class,
+                                                           registry::toImpl);
+
+        Counter counter1 = store.getOrRegisterMetric(meta1, Counter.class, TAGS);
+        Counter counter2 = store.getOrRegisterMetric(meta1, Counter.class, TAGS);
+        assertEquals(counter1, counter2);
+    }
+    @Test
+    void testAvoidDuplicates2() {
+        Tag[] TAGS = {new Tag("foo", "1"), new Tag("bar", "1")};
+        Metadata meta1 = Metadata.builder()
+                .withName("a")
+                .withType(MetricType.COUNTER)
+                .build();
+
+        NoOpMetricRegistry registry = NoOpMetricRegistry.create(MetricRegistry.Type.APPLICATION);
+
+        MetricStore<NoOpMetric> store = MetricStore.create(REGISTRY_SETTINGS,
+                                                           NoOpMetricRegistry.NO_OP_METRIC_FACTORIES,
+                                                           null,
+                                                           null,
+                                                           MetricRegistry.Type.APPLICATION,
+                                                           NoOpMetric.class,
+                                                           registry::toImpl);
+
+        Counter counter1 = store.getOrRegisterMetric(meta1, Counter.class, TAGS);
+        Counter counter2 = store.getOrRegisterMetric(meta1, Counter.class, TAGS);
+        assertEquals(counter1, counter2);
+    }
+    @Test
+    void testAvoidDuplicates3() {
+        Tag[] TAGS = {new Tag("foo", "1"), new Tag("bar", "1"), new Tag("baz", "1")};
+        Metadata meta1 = Metadata.builder()
+                .withName("a")
+                .withType(MetricType.COUNTER)
+                .build();
+
+        NoOpMetricRegistry registry = NoOpMetricRegistry.create(MetricRegistry.Type.APPLICATION);
+
+        MetricStore<NoOpMetric> store = MetricStore.create(REGISTRY_SETTINGS,
+                                                           NoOpMetricRegistry.NO_OP_METRIC_FACTORIES,
+                                                           null,
+                                                           null,
+                                                           MetricRegistry.Type.APPLICATION,
+                                                           NoOpMetric.class,
+                                                           registry::toImpl);
+
+        Counter counter1 = store.getOrRegisterMetric(meta1, Counter.class, TAGS);
+        Counter counter2 = store.getOrRegisterMetric(meta1, Counter.class, TAGS);
+        assertEquals(counter1, counter2);
+    }
 }
```

**Comment by tjquinno** [comment]
I assume, because your tests specifically instantiate a no-op registry, that you do _not_ see the error with the full-featured implementation?

**Comment by vgovtva** [comment]
The tests above are the simplest way I could reproduce the issue without spending too much effort setting it up, and isolating the cause.
I literally searched for unit tests which used `Tag` object, and copy pasted it a couple of times. So I don't actually know what "no-op registry means", I'm just a naïve user of the framework.

The issue does exist in the full-feature implementation. We noticed this issue when updating from Helidion 1.4.4 (took a while to get around updated major version) to 3.0.2, and our metrics broke. Or more specifically, counters stopped incrementing.
In practice our code looks something like this (Not a working example but shows the gist):
```
@ApplicationScoped
public class ResolverResponsesCounter {

   @Inject
    private final MetricRegistry metricRegistry;
    static final String metricName = "OurCounter";
    
    void incrementOnRequest(String dimensionValue1, String dimensionValue2) {
      Metadata metadata = getOrSetMetadata(metricName);

      metricRegistry.counter(
          metadata,
          new Tag("gqlPath", dimensionValue1),
          new Tag("gqlErrorType", dimensionValue2)
       ).inc();
    }
```

After investigating the root cause, we found out that swapping the tags would fix the issue.

P.S. Java is not my native language, and I'm not sure if I get the full meaning behind the terms you use, so apologies for that.
